common/src/main/java/com/wynntils/features/user/CombatXPGainMessageFeature.java

```diff
@@ -0,0 +1,101 @@
+import com.wynntils.wynn.utils.WynnUtils;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+
+public class CombatXPGainMessageFeature extends UserFeature {
+
+    @Config
+    public int tickDelay = 20;
+
+    private long tickCounter = 0;
+
+    private float newTickXP = 0;
+    private float lastTickXP = 0;
+    private float trackedPercentage = 0;
+    private int trackedLevel = 0;
+
+    @SubscribeEvent
+    public void onWorldStateChange(WorldStateEvent event) {
+        tickCounter = 0;
+        newTickXP = 0;
+        lastTickXP = 0;
+        trackedPercentage = 0;
+    }
+
+    @SubscribeEvent
+    public void onTick(ClientTickEvent.End event) {
+        if (!WynnUtils.onWorld()) return;
+
+        tickCounter++;
+
+        if (tickCounter % tickDelay != 0) return;
+
+        tickCounter = 0;
+
+        CharacterManager.CharacterInfo data = WynnUtils.getCharacterInfo();
+
+        int newLevel = data.getXpLevel();
+
+        // You get division by zero errors when you're at the level cap (i.e. 106 in Wynncraft 2.0.1).
+        // This needs to be updated if the level cap is ever raised.
+        if (newLevel > 105) return;
+
+        if (trackedLevel == 0) {
+            trackedLevel = newLevel;
+        }
+
+        // Handle levelling up in an active session, otherwise you might see a message like "+500 XP (-90.37%)"
+        if (newLevel != trackedLevel) {
+            trackedLevel = newLevel;
+            lastTickXP = 0;
+            trackedPercentage = 0;
+        }
+
+        newTickXP = data.getCurrentXp();
+
+        if (newTickXP == lastTickXP) return;
+
+        int neededXP = data.getXpPointsNeededToLevelUp();
+        // The purpose of this if/else statement here is to account for the case when a player joins a
+        // Wynncraft world and the lastTickXP variable is still 0, because it hasn't been updated yet.
+        // Wynncraft will send the saved XP points to the player within a few ticks of joining the world,
+        // but lastTickXP will still equal 0. So, what we do here to ensure that we don't get some message
+        // like "+56403 XP (0.0%)" is to just gather the percentage from the newTickXP variable instead.
+        // Thus giving us a message like "+56403 XP (42.7%)".
+        if (lastTickXP != 0) {
+            trackedPercentage = lastTickXP / neededXP;
+        } else {
+            trackedPercentage = newTickXP / neededXP;
+        }
+
+        float gainedXP = newTickXP - lastTickXP;
+
+        float percentGained = gainedXP / neededXP;
+
+        float percentChange;
+        // The reason we do this is for the same reason as the if/else statement above. You can get cases where
+        // both trackedPercentage and percentGained are equal, which would result in a (0.0%) message.
+        if (trackedPercentage != percentGained) {
+            percentChange = percentGained * 100;
+        } else {
+            percentChange = trackedPercentage * 100;
+        }
```

Comment:
My gut feeling here is that a lot of this seems like stuff that CharacterManager should be responsible for. But I can't clearly say how that should be done, so maybe it can stay here. It all seems a bit convoluted, but your comments indicate that this is since the problem is more complex than it seems, so I guess it is okay.

Reply:
It would be nice if it could be more abstract, yea- the only problem is that there are so many cases I ran into while working out the kinks of this system that there really does need to be some sort of context-dependent way to keep catching and handling odd quirks.